### PR TITLE
Schema exposure from Karl & Claire + Karl refactoring.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi Max,

Here come:

* Schema export from Claire, curl its `/.current?schema` for JSON, or `/.current?schema={fs,md}`.
* Schema export from Karl, curl its `/?schema` for JSON, or `/?schema={fs,md}`.
* Karl migrated to a single struct with parameters.
* Karl has its native `/status`, and `/status?{parameters,params,p,full,f}` returns these parameters alongside `{"up":true}`.

Tested with `sudo ./.current/test --karl_nginx_config_file /etc/nginx/sites-enabled/karl_unittest`.

Thanks,
Dima

### CODE REVIEW

File: Karl/claire.h [hunk 1]
```diff
@@ -133,9 +153,12 @@ class GenericClaire final {
...
   void Register(status_generator_t status_filler = nullptr, bool require_karls_confirmation = false) {
-    // Register this Claire with Karl and spawn the thread to send regular keepalives.
-    // If `require_karls_confirmation` is true, throw if Karl can be not be reached.
-    // If `require_karls_confirmation` is false, just start the keepalives thread.
+    // Register this Claire with Karl and spawn the thread to send regular
+    // keepalives.
+    // If `require_karls_confirmation` is true, throw if Karl can be not
+    // be reached.
+    // If `require_karls_confirmation` is false, just start the keepalives
+    // thread.
```
Comment (mzhurovich) [inline Karl/claire.h]: Something went wrong here :-/

File: Karl/claire.h [hunk 2]
```diff
@@ -133,9 +153,12 @@ class GenericClaire final {
   void RemoveDependency(const ClaireServiceKey& service) { dependencies_.erase(service); }
 
   void Register(status_generator_t status_filler = nullptr, bool require_karls_confirmation = false) {
-    // Register this Claire with Karl and spawn the thread to send regular keepalives.
-    // If `require_karls_confirmation` is true, throw if Karl can be not be reached.
-    // If `require_karls_confirmation` is false, just start the keepalives thread.
+    // Register this Claire with Karl and spawn the thread to send regular
+    // keepalives.
+    // If `require_karls_confirmation` is true, throw if Karl can be not
```
Comment (mzhurovich) [inline Karl/claire.h]: s/`can be not`/`can not`/

File: Karl/claire.h [hunk 3]
```diff
@@ -146,7 +169,8 @@ class GenericClaire final {
       if (require_karls_confirmation) {
         const std::string route = karl_keepalive_route_ + "&confirm";
         // The call to `SendKeepaliveToKarl` is blocking.
-        // With "&confirm" at the end, the call to Karl would require Karl calling Claire back.
+        // With "&confirm" at the end, the call to Karl would require Karl
+        // calling Claire back.
```
Comment (mzhurovich) [inline Karl/claire.h]: Oops, what's happened with all the comments?

File: Karl/claire.h [hunk 4]
```diff
@@ -277,7 +302,8 @@ class GenericClaire final {
     } catch (const current::Exception&) {
       last_keepalive_attempt_result_.status = KeepaliveAttemptStatus::CouldNotConnect;
     }
-    // TODO(dk+mz): Should it really throw in keepalive thread? It will crash the binary.
+    // TODO(dk+mz): Should it really throw in keepalive thread? It will
+    // crash the binary.
```
Comment (mzhurovich) [inline Karl/claire.h]: Suppose this comment should be replaced with the one stating that there is a `catch` clause.

**Comment by dkorolev** [comment]
@mzhurovich , 7dcf701 fixes [some] indents.

It's `make indent` with a broken brackets balance, which wraps comments "correctly" with the lambda and everything else being indented 50+ spaces. Don't make check and drive.

File: Karl/claire.h [hunk 5]
```diff
@@ -277,7 +296,9 @@ class GenericClaire final {
     } catch (const current::Exception&) {
       last_keepalive_attempt_result_.status = KeepaliveAttemptStatus::CouldNotConnect;
     }
-    // TODO(dk+mz): Should it really throw in keepalive thread? It will crash the binary.
+    // OK to throw here. In the thread that sends repeated keepalives, this exception would be caught,
+    // and if an exception is thrown during the initial registration, it is an emergency unless the user
+    // decides otherwise.
```
Comment (mzhurovich) [inline Karl/claire.h]: 👍

File: Karl/schema_karl.h [hunk 6]
```diff
@@ -35,35 +35,134 @@ SOFTWARE.
 namespace current {
 namespace karl {
 
+// Karl's startup parameters.
```
Comment (mzhurovich) [inline Karl/schema_karl.h]: Looks like this should be `Karl's Nginx parameters.` and `startup parameters` are declared below.

File: Karl/schema_karl.h [hunk 7]
```diff
@@ -35,35 +35,134 @@ SOFTWARE.
...
+  CURRENT_FIELD_DESCRIPTION(public_url,
+                            "The prefix of the URL to append to have the JSONs containing per-service proxies "
+                            "URLs browsable in one click.");
+  CURRENT_FIELD(svg_name, std::string, "Karl");
+  CURRENT_FIELD_DESCRIPTION(svg_name, "The top-level image name on the GraphViz-rendered SVG.");
+  CURRENT_FIELD(github_repo_url, std::string, "");
+  CURRENT_FIELD_DESCRIPTION(github_repo_url,
+                            "The top-level GitHub repo URL to have git commit hashes browsable.");
+  CURRENT_FIELD(nginx_parameters, KarlNginxParameters, KarlNginxParameters(0, ""));
+  CURRENT_FIELD_DESCRIPTION(nginx_parameters, "Nginx manager confix.");
```
Comment (mzhurovich) [inline Karl/schema_karl.h]: s/confix/config/

File: Karl/schema_karl.h (hunk 6 above)
Comment (dkorolev) [inline Karl/schema_karl.h]: :+1: bec4ba6

File: Karl/karl.h [hunk 8]
```diff
@@ -480,11 +502,38 @@ class GenericKarl final : private KarlNginxManager<ServiceStorage<SherlockStream
       } catch (const Exception&) {
         r("Karl registration error.\n", HTTPResponseCode.InternalServerError);
       }
+    }
+  }
+
+  void ServeFleetStatus(Request r) {
+    const auto& qs = r.url.query;
+    if (qs.has("schema")) {
+      using uber_t = Variant<karl_status_t, claire_status_t, persisted_keepalive_t, KarlParameters>;
```
Comment (mzhurovich) [inline Karl/karl.h]: This type name rocks!

Comment (dkorolev) [inline Karl/karl.h]: Воистину!

File: Karl/karl.h [hunk 9]
```diff
@@ -314,7 +335,8 @@ class GenericKarl final : private KarlNginxManager<ServiceStorage<SherlockStream
         // Respond with "200 OK" in any case.
         r("NOP\n");
       }
-    } else if (r.method == "POST") {
+    } else {
+      // Accepting keepalives as either POST or GET now, no harm in this. -- D.K.
```
Comment (mzhurovich) [inline Karl/karl.h]: `GET` request with non-empty body is pretty weird imho.

Comment (dkorolev) [inline Karl/karl.h]: No harm :-)

File: Karl/schema_karl.h (hunk 7 above)
Comment (dkorolev) [inline Karl/schema_karl.h]: :+1: 2692466

**Pull request merged**